File: schedule_server/utils/jobs.py
from apscheduler.triggers.cron import CronTrigger
from apscheduler.triggers.date import DateTrigger
from apscheduler.triggers.interval import IntervalTrigger


def cron_job_data(job_trigger: CronTrigger) -> dict:
    """Returns data job data.
    :param job_trigger: CronTrigger object.
    return:
    """
    job_dict = {}
    for f in job_trigger.fields:
        job_dict[f.name] = str(f)
    return job_dict


def date_job_data(job_trigger: DateTrigger) -> dict:
    """Returns data job data.
    :param job_trigger: An apscheduler.job.Job instance.
    return:
    """
    datetime = job_trigger.__str__()[5:25]
    # 2023-10-01 00:00:00
    return {
        "year": int(datetime[0:4]),
        "month": int(datetime[5:7]),
        "day": int(datetime[8:10]),
        "hour": int(datetime[11:13]),
        "minute": int(datetime[14:16]),
        "second": int(datetime[17:19])
    }


def interval_job_data(job_trigger: IntervalTrigger) -> dict:
    """Returns data job data.
    :param job_trigger: An apscheduler.job.Job instance.
    return:
    """
    interval = job_trigger.__str__()[9:-1]
    if "day" in interval:
        # 暂不支持天以上的间隔
        return {}
    time_list = interval.split(":")
    '''
    val[0:00:01] 1s
    val[0:01:00] 1m
    val[1:00:00] 1h
    '''
    return {
        "hour": int(time_list[0]),
        "minute": int(time_list[1]),
        "second": int(time_list[2])
    }


def get_job_name(job) -> str:
    """Returns job name.
    :param job: An apscheduler.job.Job instance.
    :return: task name
    """
    return job.args[0]

File: frontend/src/service/api/Jobs.ts
import { httpSingle } from '@/lib/axios'

// 定义不同类型任务的数据结构
export interface DateJobData {
  year: number
  month: number
  day: number
  hour: number
  minute: number
  second: number
}

export interface CronJobData {
  year: string | number
  month: string | number
  day: string | number
  week?: string
  day_of_week: string
  hour: string | number
  minute: string | number
  second: string | number
}

export interface IntervalJobData {
  hour: number
  minute: number
  second: number
}

// 主接口定义
export interface IJob {
  job_id: string
  name: string
  type: 'date' | 'cron' | 'interval'
  request_url: string
  next_run_time: string
  status: string
  data: DateJobData | CronJobData | IntervalJobData
}

// API 请求参数接口定义
export interface DJob {
  job_id?: string
  url?: string
  year: number
  month: number
  day: number
  hour: number
  minute: number
  second: number
}

export interface CJob {
  job_id?: string
  url?: string
  day_of_week: string
  month: string
  day: string
  hour: string
  minute: string
  second: string
}

export interface TJob {
  job_id?: string
  url?: string
  hours: number
  minutes: number
  seconds: number
}

/**
 * 指定【 baseUrl 】新增 date定时任务
 * @param mockUrl
 * @param params
 * @returns {*}
 */
export const C_DateJob = (baseUrl: string, params: DJob) => {
  return httpSingle({
    url: `${baseUrl}/scheduler/date/add_job`,
    method: 'post',
    data: params
  })
}

/**
 * 指定【 baseUrl 】新增 cron定时任务
 * @param baseUrl
 * @param params
 * @returns {*}
 */
export const C_CronJob = (baseUrl: string, params: CJob) => {
  return httpSingle({
    url: `${baseUrl}/scheduler/cron/add_job`,
    method: 'post',
    data: params
  })
}

/**
 * 指定【 baseUrl 】新增 interval定时任务
 * @param baseUrl
 * @param params
 * @returns {*}
 */
export const C_IntervalJob = (baseUrl: string, params: TJob) => {
  return httpSingle({
    url: `${baseUrl}/scheduler/interval/add_job`,
    method: 'post',
    data: params
  })
}

/**
 * 查询指定【 baseUrl 】jobs 列表数据
 * @param baseUrl
 * @returns {*}
 */
export const R_Jobs = (baseUrl: string) => {
  return httpSingle({
    url: `${baseUrl}/scheduler/get_jobs`,
    method: 'get'
  })
}

/**
 * 查询指定【 baseUrl 】中，指定【 jobID 】的 mapping 信息
 * @param baseUrl
 * @param jobID
 * @returns {*}
 */
export const R_Job = (baseUrl: string, jobID: string) => {
  return httpSingle({
    url: `${baseUrl}/scheduler/get_jobs?job_id=${jobID}`,
    method: 'get'
  })
}

/**
 * 删除指定【 baseUrl 】中，指定【 jobID 】的 mapping 信息
 * @param baseUrl
 * @param jobID
 * @returns {*}
 */
export const D_Job = (baseUrl: string, jobID: string) => {
  return httpSingle({
    url: `${baseUrl}/scheduler/remove_job?job_id=${jobID}`,
    method: 'delete'
  })
}

/**
 * 暂停指定【 baseUrl 】中，指定【 jobID 】的 mapping 信息
 * @param baseUrl
 * @param jobID
 * @returns {*}
 */
export const Pause_Job = (baseUrl: string, jobID: string) => {
  return httpSingle({
    url: `${baseUrl}/scheduler/pause_job?job_id=${jobID}`,
    method: 'put'
  })
}

/**
 * 回复指定【 baseUrl 】中，指定【 jobID 】的 mapping 信息
 * @param baseUrl
 * @param jobID
 * @returns {*}
 */
export const Resume_Job = (baseUrl: string, jobID: string) => {
  return httpSingle({
    url: `${baseUrl}/scheduler/resume_job?job_id=${jobID}`,
    method: 'put'
  })
}
